src/libraries/System.Diagnostics.Tracing/ref/System.Diagnostics.Tracing.cs

```diff
@@ -190,21 +190,28 @@ public partial class EventSource : System.IDisposable
         protected void WriteEvent(int eventId, long arg1, long arg2) { }
         protected void WriteEvent(int eventId, long arg1, long arg2, long arg3) { }
         protected void WriteEvent(int eventId, long arg1, string? arg2) { }
+        [System.Diagnostics.CodeAnalysis.RequiresUnreferencedCode("EventSource WriteEvent will serialize the whole object graph. Trimmer will not safely handle this case because properties may be trimmed. This can be suppressed if the object is a primitive type")]
         protected void WriteEvent(int eventId, params object?[] args) { }
         protected void WriteEvent(int eventId, string? arg1) { }
         protected void WriteEvent(int eventId, string? arg1, int arg2) { }
         protected void WriteEvent(int eventId, string? arg1, int arg2, int arg3) { }
         protected void WriteEvent(int eventId, string? arg1, long arg2) { }
         protected void WriteEvent(int eventId, string? arg1, string? arg2) { }
         protected void WriteEvent(int eventId, string? arg1, string? arg2, string? arg3) { }
+        [System.Diagnostics.CodeAnalysis.RequiresUnreferencedCode("EventSource WriteEvent will serialize the whole object graph. Trimmer will not safely handle this case because properties may be trimmed. This can be suppressed if the object is a primitive type")]
```

Comment:
Where does WriteEventCore hit trouble? The caller has already done the event data serialization in this case. EventData is a pointer and size, similar to the caller passing a Span\<byte>.

Reply:
It hits trouble here:

https://github.com/dotnet/runtime/blob/872d9e47d1c3fe7e4ad262340a548977c0d46ce1/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L1269-L1271

That line reflects over all the ParameterTypes in the corresponding eventId's method. If a parameter in the method uses a non-primitive type (like a custom class that has `EventDataAttribute` on it), the trimmer may trim properties out of the custom class, or properties in the object's graph.

This would then change things like `WriteMetadata`:

https://github.com/dotnet/runtime/blob/8fdcecc4dd5baa36e394e885f18ba23ffa7c6178/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/TraceLogging/InvokeTypeInfo.cs#L39-L63

Which use all the properties on the type, recursively.

Reply:
Yes, the problem is in the EventSource serializer code (TypeAnalysis ctor) and warning all the callers in the full call graph all the way up to a public or protected method. I also went through the framework event sources that consume these dangerous APIs and suppressed the warning after manually checking that they are safe